zingo-testutils/src/lib.rs

```diff
@@ -243,6 +244,272 @@ pub async fn load_wallet(
+            if note["created_in_txid"] == txid {
+                sapling_notes += 1;
+            }
+        }
+    }
+
+    if let JsonValue::Array(pending_sapling_notes) = &notes["pending_sapling_notes"] {
+        for note in pending_sapling_notes {
+            if note["created_in_txid"] == txid {
+                sapling_notes += 1;
+            }
+        }
+    }
+
+    if let JsonValue::Array(unspent_orchard_notes) = &notes["unspent_orchard_notes"] {
+        for note in unspent_orchard_notes {
+            if note["created_in_txid"] == txid {
+                orchard_notes += 1;
+            }
+        }
+    }
+
+    if let JsonValue::Array(pending_orchard_notes) = &notes["pending_orchard_notes"] {
+        for note in pending_orchard_notes {
+            if note["created_in_txid"] == txid {
+                orchard_notes += 1;
+            }
+        }
+    }
+
+    TxNotesCount {
+        transparent_tx_notes: transparent_notes,
+        sapling_tx_notes: sapling_notes,
+        orchard_tx_notes: orchard_notes,
+    }
+}
+
+/// Returns total actions for txid.
+pub async fn tx_actions(
+    sender: &LightClient,
+    recipient: Option<&LightClient>,
+    txid: &str,
+) -> TxActions {
+    let tx_ins = tx_inputs(sender, txid).await;
+    let tx_outs = if let Some(rec) = recipient {
+        tx_outputs(rec, txid).await
+    } else {
+        TxNotesCount {
+            transparent_tx_notes: 0,
+            sapling_tx_notes: 0,
+            orchard_tx_notes: 0,
+        }
+    };
+    let tx_change = tx_outputs(sender, txid).await;
+
+    let calculated_sapling_tx_actions = cmp::max(
+        tx_ins.sapling_tx_notes,
+        tx_outs.sapling_tx_notes + tx_change.sapling_tx_notes,
+    );
+    let final_sapling_tx_actions = if calculated_sapling_tx_actions == 1 {
+        2
+    } else {
+        calculated_sapling_tx_actions
+    };
+
+    let calculated_orchard_tx_actions = cmp::max(
+        tx_ins.orchard_tx_notes,
+        tx_outs.orchard_tx_notes + tx_change.orchard_tx_notes,
+    );
+    let final_orchard_tx_actions = if calculated_orchard_tx_actions == 1 {
+        2
+    } else {
+        calculated_orchard_tx_actions
+    };
+
+    TxActions {
+        transparent_tx_actions: cmp::max(
+            tx_ins.transparent_tx_notes,
+            tx_outs.transparent_tx_notes + tx_change.transparent_tx_notes,
+        ),
+        sapling_tx_actions: final_sapling_tx_actions,
+        orchard_tx_actions: final_orchard_tx_actions,
+    }
+}
+
+/// Returns the total transfer value of txid.
+pub async fn total_tx_value(client: &LightClient, txid: &str) -> u64 {
+    let notes = client.do_list_notes(true).await;
+
+    let mut tx_spend: u64 = 0;
+    let mut tx_change: u64 = 0;
+    if let JsonValue::Array(spent_utxos) = &notes["spent_utxos"] {
+        for utxo in spent_utxos {
+            if utxo["spent"] == txid || utxo["unconfirmed_spent"] == txid {
+                tx_spend += utxo["value"].as_u64().unwrap_or(0);
```

Comment:
I think this should panic if error during unwrap

```suggestion
                tx_spend += utxo["value"].as_u64().unwrap();
```

Reply:
change made in [6e3cb8b]